librz/util/subprocess.c

```diff
@@ -890,10 +902,103 @@ static void destroy_child_env(char **child_env) {
 	free(child_env);
 }
 
-RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
+static bool init_pipes(RzSubprocess *proc, const RzSubprocessOpt *opt, int stdin_pipe[], int stdout_pipe[], int stderr_pipe[], RzPty *new_pty) {
+	RzPty *pty = opt->pty;
+
+	/* If we need to use a PTY, we should create one right now */
+	if (!pty && (opt->stdin_pipe == RZ_SUBPROCESS_PIPE_PTY || opt->stdout_pipe == RZ_SUBPROCESS_PIPE_PTY || opt->stderr_pipe == RZ_SUBPROCESS_PIPE_PTY)) {
+		pty = rz_subprocess_openpty(NULL, NULL, NULL);
+		if (!pty) {
+			return false;
+		}
+		new_pty = pty;
+	}
+
+	if (pty) {
+		proc->master_fd = pty->master_fd;
+		proc->slave_fd = pty->slave_fd;
+	}
+
+	switch (opt->stdin_pipe) {
+	case RZ_SUBPROCESS_PIPE_CREATE:
+		if (rz_sys_pipe(stdin_pipe, true) == -1) {
+			perror("pipe");
+			return false;
+		}
+		proc->stdin_fd = stdin_pipe[1];
+		break;
+	case RZ_SUBPROCESS_PIPE_PTY:
+		stdin_pipe[0] = pty->slave_fd;
+		stdin_pipe[1] = pty->master_fd;
+		proc->stdin_fd = stdin_pipe[1];
+		break;
+	case RZ_SUBPROCESS_PIPE_STDOUT:
+		RZ_LOG_ERROR("Invalid pipe option 'RZ_SUBPROCESS_PIPE_STDOUT' for stdin. Ignoring...\n");
+		break;
+	case RZ_SUBPROCESS_PIPE_NONE:
+		break;
+	}
```

Comment:
extract in a function

Reply:
I think is better the way it is right now. Extracting it to a function makes the code very convoluted and contrived since in every case, we are doing something different. Also, the reduction in code size is not significant.